### PR TITLE
External layer support

### PR DESCRIPTION
**Pull request by tbennun**
This PR introduces support for externally-compiled layers. The layer configuration points to a file and an optional layer name, and LBANN will dynamically load and initialize it at setup time.

The PR also includes an example external layer, some infrastructure (some of which may no longer be necessary), and a fix to include paths. Merging #2175 and #2176 will simplify this PR further.

@benson31 @bvanessen 

Please squash when merging.

### CODE REVIEW

**Comment by bvanessen** [comment]
@tbennun Please add an entry to the ReleaseNotes.txt file

**Comment by tbennun** [comment]
@bvanessen done, also cleaned up the PR.

**Review by bvanessen — APPROVED** [review]

File: src/layers/misc/external.cpp [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+#include "lbann/utils/sync_info_helpers.hpp"
+
+#include <algorithm>
+#include <cstdio>
+#include <dlfcn.h>
+
+namespace lbann {
+
+external_layer_setup_t load_external_library(const std::string& filename,
+                                             std::string layer_name)
```
Comment (benson31) [inline src/layers/misc/external.cpp:37]:
```suggestion
                                             std::string const& layer_name)
```

File: src/layers/misc/external.cpp [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+      dlerror());
+    return nullptr;
+  }
+
+  ////////////////////////////////////////////////////
+  // Collect functions from libraries based on device
+  if (layer_name.length() == 0)
+    layer_name = "layer";
+
+  std::string setup_funcname = std::string("setup_") + layer_name;
```
Comment (benson31) [inline src/layers/misc/external.cpp:56]:
```suggestion
  std::string const setup_funcname = std::string("setup_") + layer_name;
```

File: include/lbann/layers/misc/external.hpp [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+                                         El::Device device,
+                                         lbann_comm* comm);
+
+/** @brief Create layer from an external library
+ *
+ *  Expects any number of input tensors. Invokes a shared object (e.g., .so
+ *  file) to call the layer.
+ */
+external_layer_setup_t load_external_library(const std::string& filename,
+                                             std::string layer_name);
```
Comment (benson31) [inline include/lbann/layers/misc/external.hpp:48]:
```suggestion
                                             std::string const& layer_name);
```

**Review by benson31 — APPROVED** [review]
LGTM

File: docs/layers/miscellaneous_layers.rst [hunk 4]
```diff
@@ -181,6 +182,45 @@ Arguments:
...
+
+
+.. note:: An example layer can be found in ``src/layers/unit_test/example_layer.cpp``.
+
+Arguments:
+
+   :filename: (``string``) Library file name or path.
+
+   :layer_name: (``string``) Optional layer name for setup function 
+   (if not given, the name "layer" is used).
```
Comment (benson31) [inline docs/layers/miscellaneous_layers.rst:218]:
```suggestion
   :layer_name: (``string``) Layer name for setup function.
```

**Review by benson31 — APPROVED** [review]

**Pull request merged**